Compiler optimization fixes (QA)

GoGoOtaku opened the pull request:
This fixes two kinds of issues: Two One-Definition-Rule (ODR) violations and one out-of-bound (OOB) array access.

ODR errors can cause issues when processing link-time-optimization. Since multiple translation units contain the same symbol but those symbols don't match. This is undefined behavior and while complex linkers will probably still do the right thing (big note on probably), it can lead to runtime errors.

Specifically this was used with `typedef enum { } mapnames` which never get accessed through this symbol anyway (removing `mapnames` is a GNU extension tho and while supported by GCC, Clang and MSVC is thus not standard).
I thus just renamed it to whatever map set it is.

The other ODR violation is `SDL_scancodes`. Now this is technically not even a "different" `typedef enum` just that SDL2 added to it and now they differ. I have not found a single reason why this was copied into the C file in the first place.
Seems to me like that if SDL2 is used anyway it should just be used throughout the project.
Just like every other instance I did use local includes ("" instead of <>)  as I assume this has something to do with the Windows build process and didn't want to dig too deep into this.

The last change is the out-of-bounds access. In the current code the [256][3] 2D array is getting accessed with [0][number between 0 and 767]. This while in theory safe to do is technically out-of-bound access. This by itself isn't a big issue however aggressive loop optimization can lead to undefined behavior. This is obviously besides that accessing the data like this makes the whole 2D array thing pointless.
Now I do admit that this might not be the most elegant solution but since palettes are pointers in the rest of the code  storing the palette as a 2D array brings some issues. It leaves us with a lot of pointer arithmetics or extracting the pointer out of the 2D array anyway. Alternatively one could change the palette to be a 2D array everywhere in the code but that would cause a lot of rewrites for not very much benefit.

Blzut3 commented:
In regards to the mapnames change, I think there might be a misunderstanding.  ReflectionHLE can either be compiled for individual games as C, or can be compiled as C++ as a unified binary.  In the latter form each game is put into a namespace.  You can see the macros `REFKEEN_NS_B` and `REFKEEN_NS_E` in the relevant files.  So the qualified name for the mapnames enum is different for each game already.

I too am not sure what the motivation for copying that SDL enum was, so waiting on NY to respond to that.

I need to take a closer look at the palette change.  We may have failed to properly document it, but one of NY's philosophies with this port is to keep the code as close as possible to vanilla's (i.e. minimized diff).  Which is why you'll see the original code in comments where things had to be changed.  Obviously if the C/C++ standard is being violated/actual portability concern then it should be fixed regardless, but just FYI there may be resistance if the change is just to eliminate a code smell.

Blzut3 commented:
Looking at C23 section 6.5.3.2 paragraph 2, as well as C++23 9.3.4.5 paragraph 9, I don't think there's actually any undefined behavior in the original code.  In particular `E1[E2]` is just syntactic sugar for `*(E1 + E2)`.  However, you are correct that despite that GCC does warn about this with optimizations turned on.  Playing around with compiler explorer it appears GCC also doesn't perform its usual optimizations when this case gets hit.  Interestingly while Clang doesn't care, MSVC++ doesn't optimize an `*(E1 + E2)` variant.

Given that, it probably does make sense to make some change to the code to keep the compilers happy.  The question becomes if the proposed solution is the way to go.  Assigning `&palette1[0][0]` to a local variable could minimize the changes necessary.  It's also possible to do something like `((id0_byte_t*)palette1)[j]` and keep all the compilers happy but I can't say I'm a fan of adding casts.

GoGoOtaku commented:
As an info: I got these warnings while compiling it without any options set. Which turns out to be single executable.

Looking into my binary I can see the C++ mangled symbols although I can not see the mapnames as they probably got cut (as stated: Nothing references the mapnames enum as a whole). My assumption is this is how GCC LTO handles typedef enums as those wouldn't be an addressable symbol anyway and it probably also needs to keep it unmangled for possible C code references. I frankly don't know enough about LTO or even compilation to know exactly why it got flagged.

I changed the palette file wide because it was only referenced in FadeIn and FadeOut. I do understand the philosophy of keeping code intact tho. I mostly just wanted to share but understand if this gets rejected. It is highly specific to aggressive optimization after all.

Blzut3 commented:
Thanks for the note on LTO, I can reproduce the warnings in question with `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON` with Ubuntu 24.10.

It only complains about mapswl6.h and mapssdm.h which points to the real issue.  These files lack the namespace macros (including mapssod.h although for whatever reason that isn't part of the warning).  Its siblings the audio and gfxv headers have them so there's probably no reason for them lacking it.

NY00123 commented:
First of all, thanks for your contributions! Before replying on them, I may start with writing (at least for people unaware) why has @Blzut3 responded first.

Following the last release of ReflectionHLE, I mentioned in multiple discussion forums that while it might not necessarily mean I won't do more with ReflectionHLE, my priorities are basically elsewhere. @Blzut3 had already gotten maintainer access earlier for the purpose of publishing macOS binaries, so it may also have other uses in case they arrive.

I've planned to respond on the PR first, but thought of letting @Blzut3 know just in case. That has eventually ended in @Blzut3 responding here, which is where we stand atm.

And now, on to the replies:

> Specifically this was used with `typedef enum { } mapnames` which never get accessed through this symbol anyway (removing `mapnames` is a GNU extension tho and while supported by GCC, Clang and MSVC is thus not standard). I thus just renamed it to whatever map set it is.

Ignoring Blzut3's more recent comment about LTO for a moment, what Blzut3 mentioned earlier about the namespaces is spot-on. Generally speaking, at least for the default C++ build of a single executable, symbols from the ported codebases should reside in namespaces. Examples are `cat3d` and `catabyss`. In the Wolfenstein 3D tree, namespaces like `wl1ap14` or `sodac14` usually represent a version or two, but the idea is the same otherwise. Hence, there should be no symbol collision.

In case you use a CMake configuration with separate executables (C or C++ build), no such namespacing is required at all.
 
> The other ODR violation is `SDL_scancodes`. Now this is technically not even a "different" `typedef enum` just that SDL2 added to it and now they differ. I have not found a single reason why this was copied into the C file in the first place. Seems to me like that if SDL2 is used anyway it should just be used throughout the project. Just like every other instance I did use local includes ("" instead of <>) as I assume this has something to do with the Windows build process and didn't want to dig too deep into this.

At the time, there was a commercial Keen Dreams source port using SDL2. It also migrated to SDL scancodes from DOS ones, thus impacting the format of KDREAMS.CFG. While ReflectionHLE was originally designed to mimic DOS versions, I still decided to add partial compatibility, hence the use of SDL scancodes.

The reason for adding a copy of the header was that I didn't want to have a direct dependency on the SDL2 library within game code (outside of the game-agnostic ReflectionHLE backend).

> The last change is the out-of-bounds access. In the current code the [256][3] 2D array is getting accessed with [0][number between 0 and 767]. This while in theory safe to do is technically out-of-bound access. This by itself isn't a big issue however aggressive loop optimization can lead to undefined behavior. This is obviously besides that accessing the data like this makes the whole 2D array thing pointless. Now I do admit that this might not be the most elegant solution but since palettes are pointers in the rest of the code storing the palette as a 2D array brings some issues. It leaves us with a lot of pointer arithmetics or extracting the pointer out of the 2D array anyway. Alternatively one could change the palette to be a 2D array everywhere in the code but that would cause a lot of rewrites for not very much benefit.

I've thought of checking a C or C++ standard reference, but looks like Blzut3 has eventually done that. Even if the code is still valid, I'm not against resolving reported compiler warnings. Given my priorities for working on the project in general, that might depend on how much changes are required.

NY00123 commented:
> I need to take a closer look at the palette change. We may have failed to properly document it, but one of NY's philosophies with this port is to keep the code as close as possible to vanilla's (i.e. minimized diff). Which is why you'll see the original code in comments where things had to be changed. Obviously if the C/C++ standard is being violated/actual portability concern then it should be fixed regardless, but just FYI there may be resistance if the change is just to eliminate a code smell.

To clarify, I've never written a proper philosophy document. I think that apart from preserving unmodified game behaviors, I assumed that preserving mostly unmodified code could help in some manner in case I'd want to make a specific general modification at a later point, like (I guess) renaming types across the whole 3D Catacombs codebase. Yes, I even went as far as preserving macros like `id0_far` as a side-effect.

> Thanks for the note on LTO, I can reproduce the warnings in question with `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON` with Ubuntu 24.10.
> 
> It only complains about mapswl6.h and mapssdm.h which points to the real issue. These files lack the namespace macros (including mapssod.h although for whatever reason that isn't part of the warning). Its siblings the audio and gfxv headers have them so there's probably no reason for them lacking it.

So there is a problem, indeed. I haven't received the warnings with a similar build on Ubuntu 20.04, but that's still an issue. The Catacomb Adventure Series' map definitions do use the namespace macros, and their `LASTMAP` values are referenced under `DebugKeys`. On the other hand, Wolfenstein 3D and derived don't seem to use the definitions at all.

Since MAPSC3D.H existed but was removed due to lack of use, I guess the same may be done under the wolfenstein 3D tree if not used. One difference is that the Catacomb 3-D sources, as released in 2014 and used for the DOS binary, don't even make any use of MAPSC3D.H, while Wolf3D's ID_HEADS.H still includes one of the headers.

Blzut3 commented:
Given the long history of source code modding for Wolf3D it's probably worthwhile to keep the headers and just add the namespace macros even if the enum is normally unused.

GoGoOtaku commented:
I can confirm that just adding the macros to the three files that lacked them does the trick.
I will also rewrite the memory thing to be localized to FadeIn function.
Lastly I would recommend to keep the removal of `SDL_Scancode` from `src/kdreams/key_scancode_converters.c` and instead add a slightly modified copy of [the original header](https://github.com/libsdl-org/SDL/blob/SDL2/include/SDL_scancode.h) (only removing `#include "SDL_stdinc.h"`) if needed. However this is currently not needed as SDL2 is a hard dependency in CMakeLists.txt.

Blzut3 commented:
With the context that NY provided my recommendation would actually be to:

- Add a comment explaining that the enum exists for support for reading the 2015 Steam release of Keen Dreams config files.  It represents the possible values that were used by that release.
- Rename the type to KD_SDL2_Scancode or something to avoid ODR complaint.  (Maybe change the prefix for the members?  Although that probably doesn't matter.)

I assume your proposal to add a copy of the header separately is to make updating more trivial, but given that the Steam version of Keen Dreams is frozen in time it would actually be counter productive to update it.  Granted, based on where SDL's comments root the values from it seems unlikely to change, but if a major version of SDL did change the values we wouldn't want to mirror that change here.

GoGoOtaku committed:
Fix ODR for mapnames typedef enum

GoGoOtaku committed:
Fix ODR for SDL_scancode typedef enum

GoGoOtaku committed:
Fix OOB array access/loop optimization

GoGoOtaku commented:
It wasn't about updating but rather about being able to build against SDL2's headers but not requiring it.
In the end I just added a comment and renamed the enum. It is a converter as you stated.

I updated the commits.

NY00123 commented:
I've had a look at the last changes, and also tried building a binary (with LTO and no debug info explicitly enabled), and I think it looks good. I might still edit the comment about the scancodes, just to clarify the enum was added for compatibility with what I referred to as the "2015 port" in ReflectionHLE. One may also find mentions of the "2015 release" in the codebase, while it's identified as "Keen Dreams 2015" in ReflectionHLE's launcher.

One more thing about the scancodes enum is that ReflectionHLE itself, or at least forks of it, don't have to use SDL. In fact, that had already occurred beforehand, with older version(s) of ReflectionHLE being ported to the Amiga. The existence of the header as a part of the Keen Dreams game tree should ensure compatibility with the 2015 data in such scenarios, even if not commonly used.

> Given the long history of source code modding for Wolf3D it's probably worthwhile to keep the headers and just add the namespace macros even if the enum is normally unused.

Well, source-modding was not a main goal of ReflectionHLE. Originally, I thought of something closer to Chocolate Doom. However, Wolfenstein 3D modding does not cover a historical DeHackEd equivalent which isn't essentially source modding. Besides, the code is out there, and as written earlier, I'm not even sure I'll change it that much in the coming years.

Blzut3 commented:
> Well, source-modding was not a main goal of ReflectionHLE.

I wouldn't say it is either, but it's not like we're going out of our way to do something here.

The current changes look good to me.  (For completeness I'll note I've been digging deeper into the 2D array as 1D question and it may be the case that the original solution was the way to make things completely defined.  Personally it feels just ambiguous enough that I'm fine continuing with the minimal solution and the compilers seem happy with it.  But I won't be too shocked if this turns into a "I told you so.")

GoGoOtaku commented:
When I initially wrote this I also thought about other solutions. Initially I wanted to rewrite FadeIn to work more like FadeOut; Without the `start  *= 3; end = end *3 +2;` However FadeIn uses the global palette which is just a `id0_byte_t id0_far` pointer and thus it would have required a lot more multiplications and I rather reduce the amount of pointer arithmetics and I am pretty sure that two `inc`s are cheaper than three `imul`s.

If I was working at ID in 1992 or whenever this was written I'd probably use a `struct { id0_byte_t red, id0_byte_t green, id0_byte_t blue } color` struct, put that in an array and most notably use that throughout the entire codebase. This isn't about rewriting Wolf3D just to flex on John Carmack tho.

NY00123 commented:
Thanks again for your contributions!
